pkg/apiserver/handlers/webhook/mutation_labels.go

```diff
@@ -0,0 +1,218 @@
+	// as a label value.
+	LabelMetadataName  = "antrea.io/metadata.name"
+	antreaSvc          = "antrea"
+	antreaSvcNamespace = "kube-system"
+)
+
+// jsonPatch contains necessary info that MutatingWebhook required
+type jsonPatch struct {
+	// Op represent the operation of this mutation
+	Op jsonPatchOperation `json:"op"`
+	// Path is a jsonPath to locate the value that need to be mutated
+	Path string `json:"path"`
+	// Value represent the value which is used in mutation
+	Value interface{} `json:"value,omitempty"`
+}
+
+func HandleMutationLabels() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		klog.V(2).Info("Received request to mutate resource labels")
+		var reqBody []byte
+		if r.Body != nil {
+			reqBody, _ = ioutil.ReadAll(r.Body)
+		}
+		if len(reqBody) == 0 {
+			klog.Errorf("Mutation webhook labelsmutator received empty request body")
+			http.Error(w, "empty request body", http.StatusBadRequest)
+			return
+		}
+		// verify the content type is accurate
+		contentType := r.Header.Get("Content-Type")
+		if contentType != "application/json" {
+			klog.Errorf("Invalid Content-Type=%s, expected application/json", contentType)
+			http.Error(w, "invalid Content-Type, expected `application/json`", http.StatusUnsupportedMediaType)
+			return
+		}
+		var admissionResponse *admv1.AdmissionResponse
+		ar := admv1.AdmissionReview{}
+		ar.TypeMeta.Kind = "AdmissionReview"
+		ar.TypeMeta.APIVersion = "admission.k8s.io/v1"
+		if err := json.Unmarshal(reqBody, &ar); err != nil {
+			klog.Errorf("Label mutation received incorrect body")
+			admissionResponse = getAdmissionResponseForErr(err)
+		} else {
+			admissionResponse = mutateResourceLabels(&ar)
+		}
+		aReview := admv1.AdmissionReview{}
+		aReview.TypeMeta.Kind = "AdmissionReview"
+		aReview.TypeMeta.APIVersion = "admission.k8s.io/v1"
+		if admissionResponse != nil {
+			aReview.Response = admissionResponse
+			if ar.Request != nil {
+				aReview.Response.UID = ar.Request.UID
+			}
+		}
+		resp, err := json.Marshal(aReview)
+		if err != nil {
+			klog.Errorf("Unable to encode response during mutation: %v", err)
+			http.Error(w, fmt.Sprintf("could not encode response: %v", err), http.StatusInternalServerError)
+		}
+		klog.V(2).Infof("Writing mutation response to MutationAdmissionHook")
+		if _, err := w.Write(resp); err != nil {
+			klog.Errorf("Unable to write response during mutation: %v", err)
+			http.Error(w, fmt.Sprintf("could not write response: %v", err), http.StatusInternalServerError)
+		}
+	}
+}
+
+// mutateResourceLabels mutates the resource labels and inserts Antrea required labels.
+func mutateResourceLabels(ar *admv1.AdmissionReview) *admv1.AdmissionResponse {
+	var result *metav1.Status
+	var patch []byte
+	var allowed bool
+	var msg string
+	patchType := admv1.PatchTypeJSONPatch
+
+	op := ar.Request.Operation
+	curRaw := ar.Request.Object.Raw
+	oldRaw := ar.Request.OldObject.Raw
+
+	switch ar.Request.Kind.Kind {
+	case "Namespace":
```

Comment:
since we only access the object's metadata, is it possible to unify the code by unmarshalling into a `metav1.ObjectMeta` object instead?

```go
var meta metav1.ObjectMeta
if err := json.Unmarshal(curRaw, &meta); err != nil {
    // ...
}
// ...
```

> To unmarshal JSON into a struct, Unmarshal matches incoming object keys to the keys used by Marshal (either the struct field name or its tag), preferring an exact match but also accepting a case-insensitive match. By default, object keys which don't have a corresponding struct field are ignored (see Decoder.DisallowUnknownFields for an alternative).

Reply:
done